enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,476 @@
+MicroShift, same as the NIC names.
+
+### Allow specific IP addresses
+This may be seen as a special case of firewalling, but using IP addresses
+instead of ports. Using the same `NetworkPolicy` resources, IPs may be
+specified to allow traffic from them and reject it from any other source.
+
+Please note this is a manual procedure for the user. Documentation will be
+added on how to perform the required actions.
+
+More information in the design details section.
+
+### Workflow Description
+**cluster admin** is a human user responsible for configuring a MicroShift
+cluster.
+
+1. The cluster admin adds specific configuration for the router prior to
+   MicroShift's start.
+2. After MicroShift started, the system will ingest the configuration and setup
+   everything according to it.
+3. The router will enabled/disabled, be exposed on the specified ports, on the
+   specified IPs, allowing only specified IP addresses, all according to the
+   cluster admin provided configuration. This includes the status in the
+   LoadBalancer service type for the router.
+
+### API Extensions
+As described in the proposal, there is an entire new section in the configuration:
+```yaml
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+  expose: # Defaults to IPs in the host. Details below.
+    hostnames:
+    - <hostname>
+    interfaces:
+    - <NIC name>
+    ipAddresses:
+    - <IP address>
+```
+
+For more information check each individual section.
+
+### Topology Considerations
+#### Hypershift / Hosted Control Planes
+N/A
+
+#### Standalone Clusters
+N/A
+
+#### Single-node Deployments or MicroShift
+Enhancement is solely intended for MicroShift.
+
+### Implementation Details/Notes/Constraints
+The default router is composed of a bunch of assets part of the MicroShift
+binary. These assets come from the rebase, copied from the original router in
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator).
+There is already a [LoadBalancer service](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/manifests/assets/router/service-cloud.yaml)
+included, which MicroShift does not copy yet. Depending on the options
+configured in MicroShift, this resource will need further customization done
+by MicroShift's start process.
+
+The rest of the changes imply only additions. The LoadBalancer controller needs
+an expansion on its capabilities to include more IPs in the status and log any
+changes to services. All the logic to control NetworkPolicy resources is also
+new.
+
+#### How config options change manifests
+Each of the configuration options described above has a direct effect on the
+manifests that MicroShift will apply after starting.
+
+The `ingress.status` option will drive whether the router `Deployment` and the
+`Service` get created.
+```yaml
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
```

Comment:
Assuming I start MicroShift with the ingress enabled: what happens to the API objects if I then edit the config to disable ingress and restart MicroShift?  Will MicroShift remove the orphaned API objects at startup?

Reply:
Only the router pod and its manifests are removed. The routes resource remains, and any user created manifests will also stay there too.
A user may well have a router different than the default one and the rest of the features need to work too. And this would not require any restarts so it needs to be there (meaning the api, controller, etc) everytime MicroShift is running.